src-utils/HydraAuctionUtils/Monads.hs

```diff
@@ -0,0 +1,110 @@
+module HydraAuctionUtils.Monads (
+  UtxoQuery (..),
+  MonadTrace (..),
+  MonadSubmitTx (..),
+  MonadQueryUtxo (..),
+  MonadNetworkId (..),
+  MonadCardanoClient,
+  logMsg,
+  submitAndAwaitTx,
+  fromPlutusAddressInMonad,
+  addressAndKeysForActor,
+) where
+
+-- Prelude imports
+import Prelude
+
+-- Cardano imports
+import Cardano.Api.UTxO qualified as UTxO
+import Hydra.Cardano.Api (
+  Address,
+  AddressInEra,
+  NetworkId,
+  PaymentKey,
+  ShelleyAddr,
+  SigningKey,
+  Tx,
+  TxIn,
+  VerificationKey,
+  fromPlutusAddress,
+  getTxId,
+  txBody,
+ )
+
+-- Plutus imports
+import Plutus.V1.Ledger.Address qualified as PlutusAddress
+
+-- HydraAuction imports
+import CardanoClient (buildAddress)
+import Control.Monad.IO.Class (MonadIO (liftIO))
+import HydraAuctionUtils.Extras.CardanoApi (networkIdToNetwork)
+import HydraAuctionUtils.Fixture (Actor, keysFor)
+
+-- MonadQueryUtxo
+
+data UtxoQuery
+  = ByAddress (Address ShelleyAddr)
+  | ByTxIns [TxIn]
+
+class Monad m => MonadQueryUtxo m where
+  queryUtxo :: UtxoQuery -> m UTxO.UTxO
+
+-- MonadNetworkId
+
+class Monad m => MonadNetworkId m where
+  askNetworkId :: m NetworkId
+
+fromPlutusAddressInMonad ::
+  MonadNetworkId m => PlutusAddress.Address -> m AddressInEra
+fromPlutusAddressInMonad address = do
+  networkId <- askNetworkId
+  let network = networkIdToNetwork networkId
+  return $
+    fromPlutusAddress network address
+
+addressAndKeysForActor ::
+  (MonadNetworkId m, MonadIO m) =>
+  Actor ->
+  m
+    ( Address ShelleyAddr
+    , VerificationKey PaymentKey
+    , SigningKey PaymentKey
+    )
+addressAndKeysForActor actor = do
+  networkId' <- askNetworkId
+  (actorVk, actorSk) <- liftIO $ keysFor actor
+  let actorAddress = buildAddress actorVk networkId'
+  pure (actorAddress, actorVk, actorSk)
+
+-- MonadTrace
+
+class Monad m => MonadTrace m where
+  type TracerMessage m = mt | mt -> m
+  stringToMessage :: String -> TracerMessage m
+  traceMessage :: TracerMessage m -> m ()
+
+logMsg :: MonadTrace m => String -> m ()
+logMsg = traceMessage . stringToMessage
+
+-- MonadSubmitTx
+
+class Monad m => MonadSubmitTx m where
+  submitTx :: Tx -> m ()
+  awaitTx :: Tx -> m ()
+
+submitAndAwaitTx :: (MonadSubmitTx m, MonadTrace m) => Tx -> m ()
+submitAndAwaitTx tx = do
+  submitTx tx
+  logMsg "Submited"
+  awaitTx tx
+  logMsg $ "Created Tx id: " <> show (getTxId $ txBody tx)
```

Comment:
this works fine if all you want to trace are `String`s but as soon as it is some other type (e.g. a structured log type) you will have `TracerMesage m ~ MyLogtype` in your context, I think it's arguably nicer to have `MonadTrace MyLogType m`

Reply:
`logMsg` is made specifically to emit strings. There is fundep type for structured log. So I guess no need for such constraints.

So for now typeclass assumes that String can be embedded to structured log type. This may be strange but makes `logMsg`, used everywhere in our code.

Reply:
oh, that's not correct, because it's injective, but yeah, it kinda shows the issue, you want to be able to trace multiple log types, the one-to-one correspondence is a bit odd, I think. okay, I think I have to think about it again

Reply:
> So for now typeclass assumes that String can be embedded to structured log type. 

it will require parsing the strings into some structured log type at some point

Reply:
> because it's injective

What is injective?

> it will require parsing the strings into some structured log type at some point

No, we just pushing strings in datatype as you may see. Any such log place could be refactored to structured log later.

Reply:
> What is injective?

the type family, what you added is an injectivity annotation

Reply:
> Any such log place could be refactored to structured log later.

no, I don't think so.